src/System.Collections/src/System/Collections/Generic/SortedList.cs

```diff
@@ -152,10 +152,31 @@ public SortedList(IDictionary<TKey, TValue> dictionary, IComparer<TKey> comparer
             if (dictionary == null)
                 throw new ArgumentNullException(nameof(dictionary));
 
-            dictionary.Keys.CopyTo(_keys, 0);
-            dictionary.Values.CopyTo(_values, 0);
-            Array.Sort<TKey, TValue>(_keys, _values, comparer);
-            _size = dictionary.Count;
+            int count = dictionary.Count;
+            if (count != 0)
+            {
+                dictionary.Keys.CopyTo(_keys, 0);
+                dictionary.Values.CopyTo(_values, 0);
+                Debug.Assert(count == _keys.Length);
+                if (count > 1)
+                {
+                    Array.Sort<TKey, TValue>(_keys, _values, comparer);
+                    comparer = Comparer; // obtain default if this is null.
+                    TKey keyPrev = _keys[0];
+                    for (int i = 1; i != _keys.Length; ++i)
+                    {
+                        TKey keyNext = _keys[i];
+                        if (comparer.Compare(keyPrev, keyNext) >= 0)
```

Comment:
I went for `>=` as "Has incorrect value". On consideration, `==` as "Has particular incorrect value we have reason to expect" is clearer as to why it is being done, so I'll change to that.